### PR TITLE
Added test that validates GC state is updated before trailing ops are processed

### PR DESCRIPTION
**Pull request by agarwal-navin**
#18758 update the logic to initialize GC state. This used to happen when container connects in "write" mode but now it happens during GC (container runtime) initialization. So, any trailing ops (or any ops for that matter) before "write" mode that use unreferenced objects will result in logging an error log.

Basically, before #18758, GC would miss logging unreferenced errors before container connected in "write" mode and now it won't. This PR adds a unit test to validate that. In the test, GC initializes with an object that is inactive, tombstone or sweep ready. This objects is then loaded, receives an op and is revived and the test validates that the appropriate events are logged.
Also, added clientId to `GCInitializationOrUpdateFailed` event and a comment that was missed in #18758.

[AB#6606](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/6606)

### CODE REVIEW

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts [hunk 1]
```diff
@@ -626,6 +629,7 @@ describe("Garbage Collection Tests", () => {
 				);
 
 				// Add reference from node 2 to node 3 and validate that revived event is logged.
+				defaultGCData.gcNodes[nodes[2]] = [nodes[3]];
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:632]: This seems like it doesn't do anything, but is it just to be consistent when "adding a reference", to update both this and call `addedOutboundReference`?

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts [hunk 2]
```diff
@@ -645,6 +649,93 @@ describe("Garbage Collection Tests", () => {
 				);
 			});
 
+			/**
+			 * Here, the base snapshot contains nodes that have timed out. The test validates that we generate errors
+			 * when these nodes are used via trailing ops. It simulates trailing ops by ensuring that garbage collector
+			 * is not loaded and GC does not run before nodeUsed is called.
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:655]: I don't understand what you mean by "ensuring that garbage collector is not loaded".  I thought the point of the last PR (and the `initalizeBaseState` call on line 687) is that GC will be loaded right away.

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts [hunk 3]
```diff
@@ -645,6 +649,93 @@ describe("Garbage Collection Tests", () => {
...
+			 */
+			it("generates events for nodes that time out on load - nodes are used via trailing ops", async () => {
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:657]: This test is a little hard to follow - it feels better suited as an e2e test where there's less in-depth manipulation of stuff.  I see several comments about simulating or doing something explicitly that other code will do in a typical setup, which feels potentially brittle.

**Review by markfields — APPROVED** [review]

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts (hunk 3 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:657]: I think this scenario is covered by the trailing ops e2e tests, so I assume your point is to get Unit Test coverage.  Maybe you can explain a little more about the benefits of this approach.

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts (hunk 1 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:632]: The reference was removed above in line 606. This adds it back. It works without this too because we reset the unreferenced timestamp in this scenario resulting the node not being in inactive+ stage. So, a revived event is logged.

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts (hunk 2 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:655]: It means GC is not loaded until "nodeUsed" is called. I will udpate the comment to make it clearer.

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts (hunk 3 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:657]: Yes, these scenarios are hard to simulate both here and in end-to-end tests. As unit tests are simpler to write, I chose to do it here.

> I see several comments about simulating or doing something explicitly that other code will do in a typical setup, which feels potentially brittle.

Isn't that what unit tests do? I agree that some of the simulation are not very apparent and we should definitely clean those up.
